### DISCUSSION TITLE
Freeform from Dev to Production

### DISCUSSION THREAD

**Discussion by kevinamezaga** [solspace/craft-freeform#121] [Questions & Answers] [answered]
Hi there, I have a question about best practices on deploying Freefrom from local dev to production.

Our workflow is deployment via GitHub to the cloud, and since we don't push user-generated data or the local database, we lose all of our forms. We could (and do) export the Freeform tables on the local server and import them on the cloud, but that's an extra step.

What are the (if any) best practices for deployment of Freeform?

**Comment by darylknight** [comment]
As you said, all the forms are user content rather than managed by Project Config. This means they should be created on the staging/production database then pulled down with all the other content, just like normal Craft entries.

**Reply by kevinamezaga** [reply]
See, but I don't (and I've seen others) not use Freeform in that way. In fact, Freeform is a backbone of my Craft web application, and without it, our services won't work.

So, if Freeform's actual *forms* could be added as config, then it would make it easier to move from Dev to Production.

**Comment by kjmartens** [comment, marked as answer]
@kevinamezaga We are actively pursuing what it might look like to have Freeform thoroughly use Project Config. The plan is to do exactly that in the coming weeks or so, but we still need to confirm everything can work this way and how we'll do it exactly. :)

EDIT: this likely may not be possible to allow for both this flow and another common flow (of continuing to allow clients to adjust forms in production). See other response below: https://github.com/solspace/craft3-freeform/discussions/121#discussioncomment-718853

We will investigate and see if we have any potential options, however. :)

**Reply by kevinamezaga** [reply]
That'd be a huge help! Currently it's a PITA to get our forms onto production when launching. Thanks.

**Comment by darylknight** [comment]
If you're considering that, please make sure Project Config for forms and fields is entirely optional. Primarily, Freeform is sold (at leaest by us) as a tool to allow clients to create and edit forms on their websites without a developer. If that suddenly becomes managed by Project Config, it's going to become a lot less useful.

**Reply by kevinamezaga** [reply]
I think at least having the option would be helpful, even if it's an option you can turn on/off. I disagree it would be less useful, it'd just be useful in a different way.

**Reply by kjmartens** [reply]
Yes, I immediately remembered this after I answered that was the primary reason we haven't added this yet. I really don't think there is a way to make it optionally allowed to be modified while the `allowAdminChanges` setting is set to false, which would make this an impossible thing to do. It would help a percentage of customers with that specific flow, but disrupt everything for likely a larger percentage of customers. :/

**Reply by kjmartens** [reply]
@kevinamezaga the problem is that it isn't optional at all. It's all or nothing from what I understand. :(

**Reply by darylknight** [reply]
Yeah we'd basically have to strip out Freeform on 40+ websites and replace it with another plugin that allowed clients to create and manage forms on their own.

**Reply by kevinamezaga** [reply]
Yeah, RIP. It's not necessary, just a thing that might be helpful to some use cases from what I'd read around.

**Reply by kjmartens** [reply]
@kevinamezaga Yes, we're very aware of a number of customers that run into this issue with this flow. And both flows are/should be valid. We may have other options to alleviate this issue for your flow Kevin, so we will investigate what alternative approaches might be viable (probably can't be project config, but perhaps some kind of migration tool, etc). :)

**Reply by kevinamezaga** [reply]
Thanks, as always, @kjmartens.
